crates/rspack_plugin_devtool/src/source_map_dev_tool_plugin.rs

```diff
@@ -341,189 +340,193 @@ impl SourceMapDevToolPlugin {
-          .unwrap_or_else(|e| panic!("{}", e.to_string()));
-        source_map_buffer
-      });
-
-      let mut asset = compilation
-        .assets()
-        .get(&filename)
-        .unwrap_or_else(|| {
-          panic!(
-            "expected to find filename '{}' in compilation.assets, but it was not present",
-            &filename
-          )
-        })
-        .clone();
-      // convert to RawSource to reduce one time source map calculation when convert to JsCompatSource
-      let raw_source = RawSource::from(code_buffer).boxed();
-      let Some(source_map_buffer) = source_map_buffer else {
-        asset.source = Some(raw_source);
-        mapped_asstes.push(MappedAsset {
-          asset: (filename, asset),
-          source_map: None,
-        });
-        continue;
-      };
-      let css_extension_detected = CSS_EXTENSION_DETECT_REGEXP.is_match(&filename);
-      let current_source_mapping_url_comment = match &self.source_mapping_url_comment {
-        Some(SourceMappingUrlComment::String(s)) => {
-          let s = if css_extension_detected {
-            URL_FORMATTING_REGEXP.replace_all(s, "\n/*$1*/")
-          } else {
-            Cow::from(s)
-          };
-          Some(SourceMappingUrlCommentRef::String(s.to_string()))
-        }
-        Some(SourceMappingUrlComment::Fn(f)) => Some(SourceMappingUrlCommentRef::Fn(f)),
-        None => None,
-      };
-
-      if let Some(source_map_filename_config) = &self.source_map_filename {
-        let chunk = file_to_chunk.get(&filename);
-        let source_type = if css_extension_detected {
-          &SourceType::Css
-        } else {
-          &SourceType::JavaScript
+    mapped_sources
+      .into_par_iter()
+      .map(|(filename, asset, source_map)| {
+        let code_buffer = {
+          let mut buffer = Vec::with_capacity(asset.size());
+          asset.to_writer(&mut buffer).into_diagnostic()?;
+          buffer
         };
-        let filename = match &self.file_context {
-          Some(file_context) => relative(Path::new(file_context), Path::new(&filename))
-            .to_string_lossy()
-            .to_string(),
-          None => filename.clone(),
+        let source_map_buffer = match source_map {
+          Some(map) => {
+            let mut buffer = Vec::new();
+            map.to_writer(&mut buffer).into_diagnostic()?;
+            Some(buffer)
+          }
+          None => None,
         };
-        let data = PathData::default().filename(&filename);
-        let data = match chunk {
-          Some(chunk) => data
-            .chunk(chunk)
-            .content_hash_optional(chunk.content_hash.get(source_type).map(|i| i.encoded())),
-          None => data,
+
+        let mut asset = compilation
+          .assets()
+          .get(&filename)
+          .unwrap_or_else(|| {
+            panic!(
+              "expected to find filename '{}' in compilation.assets, but it was not present",
+              &filename
+            )
```

Comment:
Is there a way we make this diagnostic instead of panic

Reply:
Panic is better. `filename` should always be found in `assets`. Using `unwrap_or_else` instead of `expect` provides more information when a panic occurs.